variety/Util.py

```diff
@@ -875,10 +881,50 @@ def safe_unlink(filepath):
         except Exception:
             logger.exception(lambda: "Could not delete {}, ignoring".format(filepath))
 
+    @staticmethod
+    def copy_with_replace(from_path, to_path, search_replace_map):
+        with open(from_path, "r") as file:
+            data = file.read()
+        for search, replace in search_replace_map.items():
+            data = data.replace(search, replace)
+        with open(to_path + ".partial", "w") as file:
+            file.write(data)
+            file.flush()
+        os.rename(to_path + ".partial", to_path)
+
+    @staticmethod
+    def get_exec_path():
+        return os.path.abspath(sys.argv[0])
+
 
 def on_gtk(f):
     @functools.wraps(f)
     def wrapped(*args):
         Util.add_mainloop_task(f, *args)
 
     return wrapped
+
+
+def safe_print(text, ascii_text=None, file=sys.stdout):
+    """
+    Python's print throws UnicodeEncodeError if the terminal encoding is borked. This version tries print, then logging, then printing the ascii text when one is present.
+    If does not throw exceptions even if it fails.
+    :param text: Text to print, str or unicode, possibly with non-ascii symbols in it
+    :param ascii_text: optional. Original untranslated ascii version of the text when present.
+    """
```

Comment:
I actually wonder if this is still an issue on Python 3.

(Also, would it be easier to have this method try to do the translation, and log the untranslated version if it fails?)

Reply:
Not sure regarding Python 3, I have just moved this from its previous location and haven't spent time on it. Having just one param and doing the translation in the method does look simpler.

Reply:
But this was reproducing under very strange circumstances on people's machines (e.g. some strange non-default terminal encoding), and I don't think it was Python 2's fault there - simply the printed translated text could in no way be encoded to whatever the terminal was using. So I'm sure I want to risk touching this. (though I'll definitely change its interface so a single string is passed).

Reply:
mmm, I recalled why I do it this way - the scripts for building the pot file do static analysis and actually look for _("xxx") patterns. So if we dynamically do the translation with `_(ascii_text)` in the print method, this message will actually never get into the pot file and be localized.

Reply:
basically the rule of thumb should be "do not call _() on non-literal text", if that's the case, it cannot be properly localized.